### PR TITLE
BSP-36: Add Tezos Work Queue

### PR DESCRIPTION
**Pull request by podefr**
- Add Tokenization Service to tokenize items into Tezos
- Add Work Queue to asynchronously process Tezos operations
- Switch to npm workspace repository

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,27 +4,77 @@ Item Service is an integrated service that manages gaming items for video game e
 and then tokenizes them in the blockchain.
 Items can be created, assigned to users and may evolve each at their own pace.
 
-## Run
+## Setup
 
-1. Start minilab
+Start by creating a `.env` file in this project's top level folderwith the following configuration:
```
Comment (0x103) [inline README.md:9]: If I create the `.env` file at the project's top level, the command `npm run seed-db -w item-service` doesn't work and throw the following error:
`seed: password authentication failed for user "Admin"`
I think that the environment variables are not loaded. 
If I move the `.env` in the `item-service` package, the seed command works.

File: README.md [hunk 2]
```diff
@@ -4,27 +4,77 @@ Item Service is an integrated service that manages gaming items for video game e
...
+
+```
+npm run seed-db -w item-service
+```
 
-The `http://localhost:8000/docs` url serves the OpenApi documentation.
+Deploy the Warehouse contract:
+
+```
+npm run deploy-warehouse -w item-service
```
Comment (0x103) [inline README.md:65]: Same here, If the `.env` file is at the root level, the following error occur:
`
BSP-36-add-tezos-work-queue\item-service\node_modules\@taquito\signer\dist\taquito-signer.umd.js:386
            var encrypted = key.substring(2, 3) === 'e';
                                ^
TypeError: Cannot read property 'substring' of undefined
`

File: README.md [hunk 3]
```diff
@@ -4,27 +4,77 @@ Item Service is an integrated service that manages gaming items for video game e
...
+```
+
+Notice the new contract address and paste it in the `.env` file under `WAREHOUSE_CONTRACT_ADDRESS`
+
+Start the `item-service`
+
+```
+npm run dev -w item-service
```
Comment (0x103) [inline README.md:73]: Same problem.

I Also get this error : 

 >> jetstream functionality in nats.js is preview functionality 
[1] debug:   Successfully loaded Warehouse contract at KT1CwzueGxSjihhWWCWKR2TbcuyESDpUHTtw
[1] NatsError: TIMEOUT
[1]     at Function.errorForCode (\dev\PRs\BSP-36-add-tezos-work-queue\item-service\node_modules\nats\lib\nats-base-client\error.js:98:16)
[1]     at Timeout.<anonymous> (\dev\PRs\BSP-36-add-tezos-work-queue\item-service\node_modules\nats\lib\nats-base-client\util.js:82:38)
[1]     at listOnTimeout (internal/timers.js:554:17)
[1]     at processTimers (internal/timers.js:497:7) {
[1]   code: 'TIMEOUT',
[1]   chainedError: undefined
[1] }

File: README.md [hunk 4]
```diff
@@ -4,27 +4,77 @@ Item Service is an integrated service that manages gaming items for video game e
...
+```
+
+And start the `tezos-work-queue`
+
+```
+npm run dev -w tezos-work-queue
```
Comment (0x103) [inline README.md:79]: Same problem.

And this one:

> [1]  >> jetstream functionality in nats.js is preview functionality 
[1] NatsError: TIMEOUT
[1]     at Function.errorForCode (\dev\PRs\BSP-36-add-tezos-work-queue\item-service\node_modules\nats\lib\nats-base-client\error.js:98:16)
[1]     at Timeout.<anonymous> (\dev\PRs\BSP-36-add-tezos-work-queue\item-service\node_modules\nats\lib\nats-base-client\util.js:82:38)   
[1]     at listOnTimeout (internal/timers.js:554:17)
[1]     at processTimers (internal/timers.js:497:7) {
[1]   code: 'TIMEOUT',
[1]   chainedError: undefined
[1] }

**Review by 0x103 — CHANGES_REQUESTED** [review]

File: packages/item-service/src/natsHandlers/createItem/CreateItem.ts [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
+    Logger,
+    MessageBus
+} from "common";
+
 import { Item, itemSchema } from "../../entities/item";
 import { ItemRepository } from "../../repositories/ItemRepository";
-import { Logger } from "../../common/logger/Logger";
+import { TokenizationEvents } from "../../services/tokenization/TokenizationService";
 
 export class CreateItemAirlockHandler extends AirlockHandler {
```
Comment (tkjwa) [inline packages/item-service/src/natsHandlers/createItem/CreateItem.ts:16]: We should consider adding common dependencies (logger, messagebus, ...) to the handlers to improve DX, setter injection could be acceptable for this use case

File: packages/item-service/src/services/tokenization/TokenizationService.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { Item } from "../../entities/item";
+
+export enum TokenizationEvents {
+    ITEM_CREATED = "ItemCreated"
+}
+
+export interface TokenizationService {
+    createItem(itemId: Pick<Item, "item_id">): void;
```
Comment (tkjwa) [inline packages/item-service/src/services/tokenization/TokenizationService.ts:8]: should we use a more blockchain-oriented name ? mintItem ?

File: packages/item-service/src/services/tokenization/TokenizationService.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { Item } from "../../entities/item";
+
+export enum TokenizationEvents {
+    ITEM_CREATED = "ItemCreated"
```
Comment (tkjwa) [inline packages/item-service/src/services/tokenization/TokenizationService.ts:4]: IMO this event should not live in a tokenisation context/namespace, it will still be relevant event if we'd stop writing items to blockchain

File: packages/item-service/tests/utils/mockMessageBus.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export default {
```
Comment (tkjwa) [inline packages/item-service/tests/utils/mockMessageBus.ts:1]: Could we create mocks in the test code with something like `const bus = jest.mock(MessageBus)` instead ?

File: packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import {
+    consumerOpts,
+    createInbox,
+    JetStreamClient,
+    JetStreamSubscription,
+    JSONCodec,
+    NatsConnection
+} from "nats";
+
+export class TezosWorkQueuePlugin implements RunnerPlugin {
```
Comment (tkjwa) [inline packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts:13]: Do you think we could make a new type of handler (WorkerHandler) instead of a plugin and keep a simple interface for message handling ?

File: packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+                const batchOperation = await batch.send();
+
+                await batchOperation.confirmation(1);
+
+                this.logger.debug(
+                    `Successfully processed ${JSON.stringify(
+                        operation
+                    )}`
+                );
```
Comment (tkjwa) [inline packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts:60]: what line length is the max ? could be reduced to 3 lines at least

**Review by tkjwa — APPROVED** [review]
Very nice ! lots of great additions here.

I am wondering if Tezos related code should live in the item service or not, I'll explain myself when we'll review it togehter.

Also we should merge it asap and continue improvements in other PR as it is blocking for Julien to move forward on the handlers.

Quick question about the change of folder structure: will each folder in `packages` live in its own repo ?

IMO the negative point is the use of `Pick<item_id`  construct which make the code less explicit

File: README.md (hunk 1 above)
Comment (podefr) [reply README.md:9]: can you check if you have an alias of `.env` in the `item-service` folder? the alias should be checked-in

Comment (podefr) [reply README.md:9]: should be fixed with the .env aliases added to git

File: packages/common/src/MessageBus/EventEmitterMessageBus.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  publish(eventName: string, ...args: unknown[]): void {
+    this.eventEmitter.emit(eventName, ...args);
+  }
+
+  subscribe(eventName: string, callback: (...args: unknown[]) => void): () => void {
+    this.eventEmitter.on(eventName, callback);
+
+    return () => this.eventEmitter.off(eventName, callback);
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/common/src/MessageBus/EventEmitterMessageBus.ts:18]: it's not currently used but I'm establishing now that to unsubscribe we'd just call the function returned by subscribe :) It's a pattern that I like, I think it's better than calling a `.off` method and find the eventName AND a reference to the callback again. Most pub/sub libs use this pattern now, as opposed to the one used by eventEmitter which I don't see very often anymore

File: packages/common/src/MessageBus/MessageBus.ts [hunk 12]
```diff
@@ -0,0 +1,4 @@
+export interface MessageBus {
+  publish(eventName: string, ...params: unknown[]): void;
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/common/src/MessageBus/MessageBus.ts:2]: so this got tricky. I can't infer the name from the event class, because when the class is initialized (new Class) the `.name` property can't be accessed anymore, so when I publish the event I can't get the `name` and don't know the subject name.
I had to create an Event and ensure (using TS) that the property `name` is present. Now the issue is that `name` is only there when the object is newed up, so I can't pass the class as a parameter and extract it's name when I subscribe. LOL
I thought about using a static property instead, but can't use a static property called "name" because it conflicts with...  the classes's name before it's newed up. I tried using another property name and then realized that I can't create an interface that requires a member when that member is static.

So long story short, all I could do was to ensure that the event passed to pub/sub has a property "name", set when the class is newed up, and when I subscribe to an event I use TS to verify that the event's name matches the event type. It requires passing a string to subscribe as a first argument but at least if there's a typo in the string we'll get an error :)

File: packages/item-service/src/natsHandlers/createItem/CreateItem.ts (hunk 5 above)
Comment (podefr) [reply packages/item-service/src/natsHandlers/createItem/CreateItem.ts:16]: let's make a not to chat about it and address in next PR

File: packages/item-service/src/repositories/ItemRepository.ts [hunk 13]
```diff
@@ -2,5 +2,5 @@ import { Item } from "../entities/item";
 
 export interface ItemRepository {
     addItem(item: Item): Promise<number>;
-    getItem(item_id: number): Promise<Item>;
+    getItem(item_id: Pick<Item, "item_id">): Promise<Item>;
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/item-service/src/repositories/ItemRepository.ts:5]: removed

File: packages/item-service/src/services/Tezos/TezosTokenizationService.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    async createItem(itemId: Pick<Item, "item_id">): Promise<void> {
+        const item = await this.itemRepository.getItem(itemId);
+
+        // need to remove this comment by typing WarehouseItem better
+        // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+        const warehouseItem = new WarehouseItem(item as any);
+
+        // node dependency injection doesn't support async factories
+        // so they return a Promise with the value, hence the await here.
+        // we should add a compilerPass with an async tag to work around this and maybe raise an issue with the library.
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/item-service/src/services/Tezos/TezosTokenizationService.ts:40]: yes I agree, would like to tackle this in the next PR

File: packages/item-service/src/services/tokenization/TokenizationService.ts (hunk 7 above)
Comment (podefr) [reply packages/item-service/src/services/tokenization/TokenizationService.ts:4]: fixed

File: packages/item-service/src/services/tokenization/TokenizationService.ts (hunk 6 above)
Comment (podefr) [reply packages/item-service/src/services/tokenization/TokenizationService.ts:8]: currently it's a 1:1 correspondance with the contract method, it's kind of useful :)

File: packages/item-service/tezos/deploy.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+const tezosClient = new TezosToolkit(TEZOS_RPC_URI);
+
+tezosClient.setProvider({
+    signer: new InMemorySigner(TEZOS_SECRET_KEY)
+});
+
+(async function deploy() {
+    try {
+        const operation = await tezosClient.contract.originate({
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/item-service/tezos/deploy.js:18]: unfortunatly the `seeds` folder is a convention for knex, I worry about putting tezos scripts in there because of conflicts.
I'm sure when we have to use those scripts upon deploying a new env we can find a better place for them

File: packages/tezos-work-queue/src/TezosClient.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
+import { TezosToolkit } from "@taquito/taquito";
+import { InMemorySigner } from "@taquito/signer";
+
+import { Logger } from "common";
+
+export function makeTezosClient(
```
[the remark this answers is not in the document]
Comment (podefr) [reply packages/tezos-work-queue/src/TezosClient.ts:6]: agreed, I'd like to do that in another PR or when we extract that to the common repository

File: packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts (hunk 9 above)
Comment (podefr) [reply packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts:13]: yes, a Nats jetstream handler would make a lot of sense, thanks for the suggestion. Will fix it in a next PR so we can merge this one in and make progress on the other NATS handlers

File: packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts (hunk 10 above)
Comment (podefr) [reply packages/tezos-work-queue/src/TezosWorkQueuePlugin.ts:60]: ah, I didn't do this it's set by default by prettier (it's very opinionated so we don't have to be)

File: packages/item-service/src/services/Tezos/TezosPlugin.ts [hunk 17]
```diff
@@ -43,10 +40,12 @@ export class TezosPlugin extends RunnerPlugin {
     }
 
     subscribeToEvents(): void {
-        this.messageBus.subscribe(
-            TokenizationEvents.ITEM_CREATED,
-            (itemId: unknown) =>
-                this.tezosTokenizationService.createItem(itemId as Pick<Item, "item_id">)
+        this.eventBus.subscribe(
+            "ItemCreatedEvent",
```
Comment (podefr) [inline packages/item-service/src/services/Tezos/TezosPlugin.ts:44]: no way to read this string from the constructor (the class). I pass a string instead that matches the event name, but this works since the type checking will statically verify that the 2 are indeed matching. See my comment in discussions as to why I had to resort to this solution

File: packages/common/src/MessageBus/MessageBus.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (tkjwa) [reply packages/common/src/MessageBus/MessageBus.ts]: 😲 sorry i brought that up lol
As we need to work with strings anyway we could still use an enum to reference all those events name
`class ItemCreated { name = ItemEvents.ItemCreated ....` `subscribe(ItemEvents.ItemCreated, () => ...`

**Review by tkjwa — APPROVED** [review]
good to go ! thanks for the updated. I left some conversations unresolved for later reference (the "in the next PR" ones)

**Review by 0x103 — APPROVED** [review]
